File: signalprobs.py
#s = 0
#s=int(s)
#N = float(input("Enter a number for N : "))
def sw(N):
    sw = 2*s(1-s)
    return sw
    print("switching poing is = (%f):\n", sw)
    
def signalprobs(input1sp, input2sp,input3sp,*inputsp):
    sp2AND(input1sp, input2sp);
    sp2OR(input1sp, input2sp);
    sp2XOR(input1sp, input2sp);
    sp2NAND(input1sp, input2sp);
    sp2NOR(input1sp, input2sp);
    sp3AND(input1sp, input2sp, input3sp);
    sp3OR(input1sp, input2sp, input3sp);
    sp3XOR(input1sp, input2sp, input3sp);
    sp3NAND(input1sp, input2sp, input3sp);
    sp3NOR(input1sp, input2sp, input3sp);
    spnAND(inputsp);
    spnOR(inputsp);
    spnXOR(inputsp);
    spnNAND(inputsp);
    spnNOR(inputsp);

def sp2AND(input1sp, input2sp):
    
  print("AND Gate for input probabilities (%f %f):\n",input1sp,input2sp)
  s = input1sp*input2sp;
  sw(2)

def sp2OR(input1sp, input2sp):
  print("OR Gate for input probabilities (%f %f):\n",input1sp,input2sp)
  s = 1-(1-input1sp)*(1-input2sp);
  sw(2)

def sp2XOR(input1sp, input2sp):
    print("XOR Gate for input probabilities (%f %f):\n",input1sp,input2sp)
    #s = (1-input1sp)*input2sp + input1sp*(1-input2sp)
    if (input1sp == input2sp):
        s = 0
    else:
        s = 1
    sw(2)
    
def  sp2NAND(input1sp, input2sp):
    print("NAND Gate for input probabilities (%f %f):\n",input1sp,input2sp)
    s = 1 - input1sp*input2sp;
    sw(2)
    
def  sp2NOR(input1sp, input2sp):
    print("NOR Gate for input probabilities (%f %f):\n",input1sp,input2sp)
    s = (1-input1sp)*(1-input2sp);
    sw(2)

def sp3AND(input1sp, input2sp, input3sp):
    print("AND Gate for input probabilities (%f %f %f):\n",input1sp,input2sp,input3sp)
    s = input1sp*input2sp*input3sp;
    sw(3)

def sp3OR(input1sp, input2sp, input3sp):
    print("OR Gate for input probabilities (%f %f %f):\n",input1sp,input2sp,input3sp)
    s = 1-(1-input1sp)*(1-input2sp)*(1-input3sp);
    sw(3)

def sp3XOR(input1sp, input2sp, input3sp):
    print("XOR Gate for input probabilities (%f %f %f):\n",input1sp,input2sp,input3sp)
    s = rem((input1sp + input2sp + input3sp),2)
    sw(3)

def sp3NAND(input1sp, input2sp, input3sp):
    print("NAND Gate for input probabilities (%f %f %f):\n",input1sp,input2sp,input3sp)
    s = 1 - input1sp*input2sp*input3sp;
    sw(3)
  
def sp3NOR(input1sp, input2sp, input3sp):
    print("NOR Gate for input probabilities (%f %f %f):\n",input1sp,input2sp,input3sp)
    s = (1-input1sp)*(1-input2sp)*(1-input3sp);
    sw(3)

def spnAND(*inputsp):
    #print("AND Gate for input probabilities (%f %f %f %f):\n",input1sp,input2sp,input3sp, input4sp )
    for i in range(1,N):
        if(i==1):
            s = input1sp
        else:
            s = inputsp[i] * inputsp[i-1]
    sw(N)

def spnOR(*inputsp):
    #print("OR Gate for input probabilities (%f %f %f %f):\n",input1sp,input2sp,input3sp, input4sp )
    for i in range(1,N):
        if(i==1):
            s = input1sp
        else:
            s = 1 - (1-inputsp[i]) * (1-inputsp[i-1])
    sw(N)

def spnXOR(*inputsp):
    #print("XOR Gate for input probabilities (%f %f %f %f):\n",input1sp,input2sp,input3sp, input4sp )
    for i in range(1,N):
        if(i==1):
            s = 1 - input1sp
        else:
            s = rem((inputsp[i] + inputsp[i-1]),2)
    sw(N)

def spnNAND(*inputsp):
    #print("NAND Gate for input probabilities (%f %f %f %f):\n",input1sp,input2sp,input3sp, input4sp )
    for i in range(1,N):
        if(i==1):
            s = 1 - input1sp
        else:
            s = 1 - inputsp[i] * inputsp[i-1]
    sw(N)

def spnNOR(*inputsp):
    #print("NOR Gate for input probabilities (%f %f %f %f):\n",input1sp,input2sp,input3sp, input4sp)
    for i in range(1,N):
        if(i==1):
            s = 1 - input1sp
        else:
            s = (1-inputsp[i]) * (1-inputsp[i-1])
    sw(N)

File: signalprobs.m
%%%
%%%
%%% τρέχετε το πρόγραμμα ως:
%%% signalprobs(input1sp,input2sp)
%%%
%%% Παραδείγματα:
%%% >> signalprobs(0.5,0.5)
%%% AND Gate for input probabilities (0.500000 0.500000):
%%% ans =  0.25000
%%% OR Gate for input probabilities (0.500000 0.500000):
%%% ans =  0.75000
%%% XOR Gate for input probabilities (0.500000 0.500000):
%%% NAND Gate for input probabilities (0.500000 0.500000):
%%% NOR Gate for input probabilities (0.500000 0.500000):
%%%
%%%
%%% >> signalprobs(0,0)
%%% AND Gate for input probabilities (0.00000 0.00000):
%%% ans =  0
%%% OR Gate for input probabilities (0.00000 0.00000):
%%% ans =  0
%%% XOR Gate for input probabilities (0.00000 0.00000):
%%% NAND Gate for input probabilities (0.00000 0.00000):
%%% NOR Gate for input probabilities (0.00000 0.00000):
%%%
%%% >> signalprobs(1,1)
%%% AND Gate for input probabilities (1.00000 1.00000):
%%% ans =  1
%%% OR Gate for input probabilities (1.00000 1.00000):
%%% ans =  1
%%% XOR Gate for input probabilities (1.00000 1.00000):
%%% NAND Gate for input probabilities (1.00000 1.00000):
%%% NOR Gate for input probabilities (1.00000 1.00000):
%%%
%%%
%%%
%%% Οι συναρτήσεις που υπολογίζουν τα signal probabilities 
%%% AND και OR πυλών δύο εισόδων έχουν ήδη υλοποιηθεί παρακάτω.
%%% Οι συναρτήσεις που υπολογίζουν τα signal probabilities 
%%% XOR, NAND και NOR πυλών δύο εισόδων είναι ημιτελής.
%%% (α) Σας ζητείτε να συμπληρώσετε τις υπόλοιπες ημιτελής συναρτήσεις για τον υπολογισμό
%%% των signal probabilities XOR,NAND και NOR 2 εισόδων πυλών.
%%% (β) γράψτε συναρτήσεις για τον υπολογισμό των signal probabilities 
%%% AND, OR, XOR, NAND, NOR πυλών 3 εισόδων
%%% (γ) γράψτε συναρτήσεις για τον υπολογισμό των signal probabilities 
%%% AND, OR, XOR, NAND, NOR πυλών Ν εισόδων


function s=signalprobs(input1sp, input2sp)
  sp2AND(input1sp, input2sp)
  sp2OR(input1sp, input2sp)
  
  % Οι παρακάτω συναρτήσεις πρέπει να ολοκληρωθούν για το (α)
  sp2XOR(input1sp, input2sp);
  sp2NAND(input1sp, input2sp);
  sp2NOR(input1sp, input2sp);
  
  % Οι παρακάτω συναρτήσεις πρέπει να γραφούν εξ'ολοκλήρου για το (β)
  %sp3AND(input1sp, input2sp, input3sp)
  %sp3OR(input1sp, input2sp, input3sp)
  %sp3XOR(input1sp, input2sp, input3sp);
  %sp3NAND(input1sp, input2sp, input3sp);
  %sp3NOR(input1sp, input2sp, input3sp);
  
  % Οι παρακάτω συναρτήσεις πρέπει να γραφούν εξ'ολοκλήρου για το (γ)
  %% προσοχή: πρέπει να παίζουν ανεξάρτητα του πόσες εισόδους τους δίνετε
  %spAND(input1sp, input2sp, input3sp, input4sp ...)
  %spOR(input1sp, input2sp, input3sp, input4sp ...)
  %spXOR(input1sp, input2sp, input3sp, input4sp, ...);
  %spNAND(input1sp, input2sp, input3sp, input4sp, ...);
  %spNOR(input1sp, input2sp, input3sp, input4sp, ...);
  
end
%

% 2-input AND gate truth table
% 0 0:0
% 0 1:0
% 1 0:0
% 1 1:1
%% signal probability calculator for a 2-input AND gate
%% input1sp: signal probability of first input signal
%% input2sp: signal probability of second input signal
%%        s: output signal probability
function s=sp2AND(input1sp, input2sp)
  printf("AND Gate for input probabilities (%f %f):\n",input1sp,input2sp)
  s = input1sp*input2sp;
  endfunction

% 2-input OR gate truth table
% 0 0:0
% 0 1:1
% 1 0:1
% 1 1:1
%% signal probability calculator for a 2-input OR gate
%% input1sp: signal probability of first input signal
%% input2sp: signal probability of second input signal
%%        s: output signal probability
function s=sp2OR(input1sp, input2sp)
  printf("OR Gate for input probabilities (%f %f):\n",input1sp,input2sp)
  s = 1-(1-input1sp)*(1-input2sp);
endfunction


% 2-input XOR gate truth table
% 0 0:0
% 0 1:1
% 1 0:1
% 1 1:0
%% signal probability calculator for a 2-input XOR gate
%% input1sp: signal probability of first input signal
%% input2sp: signal probability of second input signal
%%        s: output signal probability
function s=sp2XOR(input1sp, input2sp)
  printf("XOR Gate for input probabilities (%f %f):\n",input1sp,input2sp)
  if (input1sp == input2sp)
	s = 0
  else
	s = 1
  endif
  %s = ??;
endfunction


% 2-input NAND gate truth table
% 0 0:1
% 0 1:1
% 1 0:1
% 1 1:0
%% signal probability calculator for a 2-input XOR gate
%% input1sp: signal probability of first input signal
%% input2sp: signal probability of second input signal
%%        s: output signal probability
function s=sp2NAND(input1sp, input2sp)
  printf("NAND Gate for input probabilities (%f %f):\n",input1sp,input2sp)
  s = 1 - input1sp*input2sp;
  %s = ??;
endfunction



% 2-input NOR gate truth table
% 0 0:1
% 0 1:0
% 1 0:0
% 1 1:0
%% signal probability calculator for a 2-input NOR gate
%% input1sp: signal probability of first input signal
%% input2sp: signal probability of second input signal
%%        s: output signal probability
function s=sp2NOR(input1sp, input2sp)
  printf("NOR Gate for input probabilities (%f %f):\n",input1sp,input2sp)
  s = (1-input1sp)*(1-input2sp);
  %s = ??;
endfunction

function s=sp3AND(input1sp, input2sp, input3sp)
  printf("AND Gate for input probabilities (%f %f %f):\n",input1sp,input2sp,input3sp)
  s = input1sp*input2sp*input3sp;
endfunction

function s=sp3OR(input1sp, input2sp, input3sp)
  printf("OR Gate for input probabilities (%f %f %f):\n",input1sp,input2sp,input3sp)
  s = 1-(1-input1sp)*(1-input2sp)*(1-input3sp);
endfunction

function s=sp3XOR(input1sp, input2sp, input3sp)
  printf("XOR Gate for input probabilities (%f %f %f):\n",input1sp,input2sp,input3sp)
  s = rem((input1sp + input2sp + input3sp),2)
endfunction

function s=sp3NAND(input1sp, input2sp, input3sp)
  printf("NAND Gate for input probabilities (%f %f %f):\n",input1sp,input2sp,input3sp)
  s = 1 - input1sp*input2sp*input3sp;
endfunction
  
function s=sp3NOR(input1sp, input2sp, input3sp)
  printf("NOR Gate for input probabilities (%f %f %f):\n",input1sp,input2sp,input3sp)
  s = (1-input1sp)*(1-input2sp)*(1-input3sp);
endfunction

function s=spAND(input1sp, input2sp, input3sp, input4sp ...)
  %printf("AND Gate for input probabilities (%f %f %f %f):\n",input1sp,input2sp,input3sp, input4sp )
  for i = 1:N
	if(i==1)
		s = input1sp
	else
		s = input[i]sp * input[i-1]sp
	end
  end
endfunction

function s=spOR(input1sp, input2sp, input3sp, input4sp ...)
  %printf("OR Gate for input probabilities (%f %f %f %f):\n",input1sp,input2sp,input3sp, input4sp )
  for i = 1:N
	if(i==1)
		s = input1sp
	else
		s = 1 - (1-input[i]sp) * (1-input[i-1]sp)
	endif
  end
endfunction

function s=spXOR(input1sp, input2sp, input3sp, input4sp ...)
  %printf("XOR Gate for input probabilities (%f %f %f %f):\n",input1sp,input2sp,input3sp, input4sp )
  for i = 1:N
	if(i==1)
		s = 1 - input1sp
	else
		s = rem((input[i]sp + input[i-1]sp),2)
	endif
  end
endfunction

function s=spNAND(input1sp, input2sp, input3sp, input4sp ...)
  %printf("NAND Gate for input probabilities (%f %f %f %f):\n",input1sp,input2sp,input3sp, input4sp )
  for i = 1:N
	if(i==1)
		s = 1 - input1sp
	else
		s = 1 - input[i]sp * input[i-1]sp
	end
  end
endfunction

function s=spNOR(input1sp, input2sp, input3sp, input4sp ...)
  %printf("NOR Gate for input probabilities (%f %f %f %f):\n",input1sp,input2sp,input3sp, input4sp)
  for i = 1:N
	if(i==1)
		s = 1 - input1sp
	else
		s = (1-input[i]sp) * (1-input[i-1]sp)
	endif
  end
endfunction